storage/oasis/nodeapi/file/consensus.go

```diff
@@ -16,14 +16,20 @@ import (
 // Since FileConsensusApiLite is backed by a file containing the cached responses
 // to `ConsensusApiLite` calls, this data is inherently compatible with the
 // current Nexus and can thus handle heights from both Cobalt/Damask.
+//
+// If cacheVolatileEndpoints is `true`, FileConsensusApiLite will cache responses
+// from volatile endpoints and return the cached data on subsequent queries.
+// This means that the data fed to Nexus will always be consistent, but may be outdated,
+// which is acceptable in testing scenarios.
 type FileConsensusApiLite struct {
-	db           KVStore
-	consensusApi nodeapi.ConsensusApiLite
+	db                     KVStore
+	consensusApi           nodeapi.ConsensusApiLite
+	cacheVolatileEndpoints bool
```

Comment:
Uhh ... I don't have a specific scenario to point to, but I expect pain and suffering lay down this path. At best, every time we recreate the e2e cache we'll have to manually decide which changes in outputs are expected. If that happens often, it's annoying; if it happens rarely, it's surprising.
Implementing the flag also requires quite a few little changes in the codebase, which is a smell given that the flag is used for testing only.

We'll have a more stable situation if we just ignore the (very few) volatile fields in e2e tests. Example for the already-volatile `latest_update` field is here: https://github.com/oasisprotocol/nexus/blob/5cdf94ebc0cfa704a5f0efb4d91a2655446ab327/tests/e2e_regression/run.sh#L114-L114
It does slightly decrease the power of the test, but not much.

Reply:
> every time we recreate the e2e cache we'll have to manually decide which changes in outputs are expected

Yeah, but I'm hoping that this is a rare enough occurrence that it'll still be palatable. If recent memory serves, I haven't had to do it too often 

I think overall the tradeoff is worth it since we can start including other analyzers (like evm_tokens + evm_contracts) that depend on volatile endpoints in the e2e regression tests. But open to other suggestions as well that might be more elegant 😅

Reply:
Sorry, I still strongly think that caching volatile endpoints is asking for trouble, and we shouldn't do it.

It's easy enough enough to contaminate your local cache (or, god forbid, a production cache) by using the wrong flag accidentally, during testing or such. A single short run suffices, and there's no undoing it (unless we write new tools.) Once the cache is contaminated, there will be no indication of it, and no immediate failure, just hard-to-explain incorrect results that we discover possibly days later. As a similar pattern: Remember how we had false bug reports that stemmed from the incorrect caching of genesis documents (cobalt and damask were not distinguishable). The root case was quite non-obvious because the cache is so opaque.

Cache invalidation is a famously hard problem that's prone to bugs. We have the luxury of being able to have a cache that never goes stale, if we only make small accommodations elsewhere. Let's hold onto that good fortune.

To your other point: Adding other analyzers to e2e will require some thought/hacks/compromises, yes. But I actually don't expect trouble from the analyzers that pull data from the node (like `evm_tokens` and `evm_contract_code`). They fetch data from at the latest block-analyzer height, not at the latest node height; this was implemented that way partially with e2e tests in mind. So the analyzers will be hitting non-volatile endpoints. (*)
In e2e, the plan is to run those analyzers _after_ the block analyzers finish; this way, they will always fetch data at the same height. If we ran them in parallel with block analyzers, they'd still be hitting non-volatile endpoints, but at unpredictable (and thus uncacheable) heights. 

Non-node analyzers are the bigger problem: Sourcify, CoingGecko, metadata registry. At first, we'll just not include them in e2e tests. Then later we can think of mocking those endpoints (and the mock can even be cache-based).

(*) The special analyzer from this PR is a necessary exception, but is likely to remain a one-off.

Reply:
> easy enough enough to contaminate your local cache (or, god forbid, a production cache) 

Good points, agreed that it's not worth the risk of that happening. Updated accordingly

> They fetch data from at the latest block-analyzer height,

Oh nice, this will make expanding the tests way easier